### ISSUE TITLE
Support for Bitbucket's new build status API

### ISSUE THREAD

**Issue by tnguyen14** [nishio-dens/bitbucket-pullrequest-builder-plugin#65]
Bitbucket recently added the build status API for pull requests and branches http://blog.bitbucket.org/2015/11/18/introducing-the-build-status-api-for-bitbucket-cloud/

Could we update this plugin to updating the PR build status instead of leaving comments? The API doc is here https://confluence.atlassian.com/bitbucket/buildstatus-resource-779295267.html

**Comment by jdeniau** [comment]
:+1:

**Comment by middric** [comment]
:+1:

**Comment by bskart** [comment]
:+1:

**Comment by sbehn** [comment]
:+1:

**Comment by fonsecas72** [comment]
:+1:

**Comment by adam-knights** [comment]
In addition to using the new API to post success and failure it would also be great to post in progress at the start of the build.

`INPROGRESS indicates that a build for the commit is in progress but not yet complete.`

**Comment by vladistan** [comment]
:+1:

**Comment by brendanator** [comment]
:+1:

**Comment by ximenean** [comment]
:+1:

**Comment by Dahlgren** [comment]
I'm almost done with implementing support for the new Build State API. Is there any point in keeping the current method of posting a comment to the activity feed? The Build State includes both a short description and a link to the build job.

**Comment by brendanator** [comment]
@Dahlgren I would definitely remove the commenting functionality. It just ends up spamming the pull request with unnecessary comments

**Comment by Dahlgren** [comment]
I agree. In the current implementation the comments are removed.

**Comment by mvanbaak** [comment]
@Dahlgren But does it still 'approve' the PR when the build passes?

**Comment by Dahlgren** [comment]
If you by 'approve' mean the nice green tick on the avatar it will have the same behaviour as before. It's a setting which can be toggled per Jenkins job. So by default it won't but you can opt in to have the same behaviour like the 'approve' button for end users. Nothing is changed in regards to this.

**Comment by mvanbaak** [comment]
Thanks. That's exactly what I meant. We use that in our workflow now (since the buildstatus api was not in bitbucket cloud).
We wont need it anymore when this PR is merged as we will have the buildstatus updated, but people are pretty used to it to check if jenkins approved their PR :)

Nice work. Can't wait to start using it! Thanks.

**Comment by mvanbaak** [comment]
Thank you! it works perfectly!

**Comment by tnguyen14** [comment]
I just updated to 1.4.8, but I don't see the PR being updated with the build status. Is there any extra configuration I need to do?

**Comment by jdeniau** [comment]
Personally, I had to re-edit my configuration in Jenkins and change `CI Identifier` to something unique, and then it worked fine

**Comment by tnguyen14** [comment]
@jdeniau weird, but that did it for me as well. @nishio-dens does this sound like a bug to you, or is it how it's supposed to work?

**Comment by tnguyen14** [comment]
Also, it seems like the new behavior is that new builds are being constantly triggered if the PR is failing. Previously it only ran once, and even if it fails, it will only rerun when there's a code change. Is anyone else seeing this?

**Comment by andersnorgaard** [comment]
@tnguyen14 The same thing (constant rebuilds) happened for us when we upgraded today. My colleague fixed it somehow - I'll ask him what he did.

**Comment by tnguyen14** [comment]
@andersnorgaard please do! this is hogging a lot of our Jenkins resource and I feel like I'm sitting on needles.

**Comment by mvanbaak** [comment]
@tnguyen14 I had the same thing. Reconfiguring every Bitbucket PR job and putting something unique in the CI Identifier field fixed it for me.

**Comment by adam-knights** [comment]
The continuous rebuild issue is reported as https://github.com/nishio-dens/bitbucket-pullrequest-builder-plugin/issues/62 - we aren't using this plugin because of that issue. If anyone finds a good workaround it would be handy to post it against that.

**Comment by andersnorgaard** [comment]
@tnguyen14 I belive he just changed the url to no longer substitute git@bitbucket.org:${repositoryOwner}/${repositoryName}.git but be explicit git@bitbucket.org:our-org/therepo.git 

So it was probably not the specific change, but just that he poked the job...?

**Comment by ximenean** [comment]
I found that changing the CI identifier had no effect. What I did notice was that after removing any old "test this please" comments from the PR the rebuilds stopped.

**Comment by Dahlgren** [comment]
Yeah, any remaining "test this please" will trigger a rebuild

**Comment by tnguyen14** [comment]
@Dahlgren is it just old "test this please" (from before the plugin update) or any of those comment? Does this mean we can no longer use "test this please"?

**Comment by Dahlgren** [comment]
As long as there is a "test this please" on an open PR it will trigger a rebuild

**Comment by tnguyen14** [comment]
Before this update, the behavior is that this comment will only trigger a rebuild once. Even if that rebuild fails, it will not run again. Is that no longer the behavior now?

**Comment by jwismar** [comment]
The behavior used to be that the plugin parsed the comments, newest to
oldest, and if there were no "Build Started" comments more recent than the
"test this please" then it would trigger a build. If that logic is still in
place, but we're no longer adding comments upon build start/completion,
then "test this please" is always going to be the most recent comment that
the plugin looks for.

---------------------------------
John Wismar
jwismar@gmail.com

**Comment by tnguyen14** [comment]
Thanks @jwismar that is very informative. Sounds like this logic should be updated to check against an actual build on Jenkins instead of the "Build Started" comment?

**Comment by jwismar** [comment]
The symptoms look similar to issue #62, but #62 existed before the build notification API was introduced, and isn't necessarily repeatable. (We have one job out of about 40 that constantly rebuilds, and haven't updated to use the notification API yet.) With the Notification API change, the constant rebuilding was completely predictable, should have been caught before being merged, and is easy to replicate, so it should be easy for someone to fix. Issue 62 is more complicated.

**Comment by tnguyen14** [comment]
Thanks for the clarification @jwismar. I wasn't aware of that nuance.

**Comment by pdjtrifork** [comment]
The problem with the 'test this please' causing continuous rebuilds is still very present...have downgraded to 1.4.7 for the same reason :-/ Have filed an issue at https://issues.jenkins-ci.org/browse/JENKINS-32097 but fear this might not be the correct spot?

**Comment by Dahlgren** [comment]
I could submit a PR with 'test this please' behaviour removed but I'm not sure it's the desired solution?

**Comment by pdjtrifork** [comment]
@Dahlgren - Not quite, if so we could just tell users to refrain from using that feature. Was considering this solution myself:
Have the plugin add a comment to the PR, saying that the 'adhoc build request' command has been completed (or some other hard coded comment that can be easily matched by the plugin). Though this comment in the PR add no value, this seems to be the only way to maintain state. If possible the plugin should just add a comment the first time around and then update this on subsequent build requests thus moving it to the top of the comment pile chronologically.
This way the 'test this please' feature would still work but unless used explicitely no comments were added to the PR by the plugin...win-win

**Comment by Dahlgren** [comment]
Yeah, I guess we could add the comments as before if the build is triggered by a 'test this please'. It's not enough with a comment on successful build since we also need to mark builds that are in progress. This was the previous behaviour so it shouldn't be an issue to do.

TL;DR; I'll add the comments back for 'test this please' builds

**Comment by mvanbaak** [comment]
Cant you add them as reply to the original 'test this please' comment?

**Comment by kadaan** [comment]
What about just deleting the "test this please" command as soon as it is kicked off?

**Comment by pdjtrifork** [comment]
@Dahlgren point taken ;-) I agree, the comments could be used as previously but exclusively for signalling the current state of a 'test this please' request.

@kadaan that would be impossible because of the security restrictions imposed by bitbucket (cannot alter another users comments)
